src/dbnode/persist/fs/wide/entry_checksum_mismatch_checker.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package wide
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/m3db/m3/src/dbnode/persist/fs/msgpack"
+	"github.com/m3db/m3/src/dbnode/persist/schema"
+	"github.com/m3db/m3/src/x/instrument"
+
+	"go.uber.org/zap"
+)
+
+type entryWithChecksum struct {
+	idChecksum int64
+	entry      schema.IndexEntry
+}
+
+type entryChecksumMismatchChecker struct {
+	blockReader  IndexChecksumBlockReader
+	mismatches   []ReadMismatch
+	strictLastID []byte
+
+	decodeOpts msgpack.DecodingOptions
+	iOpts      instrument.Options
+
+	batchIdx  int
+	exhausted bool
+	started   bool
+}
+
+// NewEntryChecksumMismatchChecker creates a new entry checksum mismatch
+// checker, backed by the given block buffer.
+// NB: index entries MUST be checked in lexicographical order by ID.
+func NewEntryChecksumMismatchChecker(
+	blockReader IndexChecksumBlockReader,
+	opts Options,
+) EntryChecksumMismatchChecker {
+	return &entryChecksumMismatchChecker{
+		blockReader: blockReader,
+		mismatches:  make([]ReadMismatch, 0, opts.BatchSize()),
+		decodeOpts:  opts.DecodingOptions(),
+		iOpts:       opts.InstrumentOptions(),
+	}
+}
+
+func entryMismatch(e entryWithChecksum) ReadMismatch {
+	return ReadMismatch{
+		Checksum:    e.idChecksum,
+		EncodedTags: e.entry.EncodedTags,
+		ID:          e.entry.ID,
+	}
+}
+
+func (c *entryChecksumMismatchChecker) entryMismatches(
+	entries ...entryWithChecksum,
+) []ReadMismatch {
+	for _, e := range entries {
+		c.mismatches = append(c.mismatches, entryMismatch(e))
+	}
+
+	return c.mismatches
+}
+
+func (c *entryChecksumMismatchChecker) indexMismatches(checksums ...int64) []ReadMismatch {
```

Comment:
I see that you never return the result of `indexMismatches` directly from `ComputeMismatchForEntry` and basically use it for the side effect. I guess this would be more clear if you would change it to
```suggestion
func (c *entryChecksumMismatchChecker) recordIndexMismatches(checksums ...int64) {
```

And in the only place where you do use the result:
```go

			mismatches := c.indexMismatches(batch.Checksums[c.batchIdx:nextBatchIdx]...)
			c.batchIdx = nextBatchIdx + 1
			return mismatches, nil
```
You could return `c.mismatches` instead.